PurchasesCoreSwift/Networking/Backend.swift

```diff
@@ -0,0 +1,664 @@
+            var extraUserInfo = [ErrorDetails.finishableKey: finishable] as [String: Any]
+            extraUserInfo.merge(subscriberAttributesErrorInfo) { _, new in new }
+            let code = maybeNumberFromError(code: response?["code"])
+            let message = response?["message"] as? String
+            let responseError = ErrorUtils.backendError(withBackendCode: code,
+                                                        backendMessage: message,
+                                                        extraUserInfo: extraUserInfo as [NSError.UserInfoKey: Any])
+            completion(maybePurchaserInfo, responseError)
+            return
+        }
+
+        completion(maybePurchaserInfo, nil)
+    }
+
+    func escapedAppUserID(appUserID: String) throws -> String {
+        let trimmedAndEscapedAppUserID = appUserID
+            .trimmingCharacters(in: .whitespacesAndNewlines)
+            .addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)!
+
+        guard trimmedAndEscapedAppUserID.count > 0 else {
+            Logger.warn("appUserID is empty")
+            throw ErrorUtils.missingAppUserIDError()
+        }
+
+        return trimmedAndEscapedAppUserID
+    }
+
+    func subscriberAttributesToDict(subscriberAttributes: SubscriberAttributeDict) -> [String: Any] {
+        var attributesByKey: [String: Any] = [:]
+        for (key, value) in subscriberAttributes {
+            attributesByKey[key] = value.asBackendDictionary()
+        }
+        return attributesByKey
+    }
+
+    func userInfoAttributes(response: [String: Any], statusCode: Int) -> [String: Any] {
+        var resultDict: [String: Any] = [:]
+
+        let isInternalServerError = statusCode >= HTTPStatusCodes.internalServerError.rawValue
+        let isNotFoundError = statusCode == HTTPStatusCodes.notFoundError.rawValue
+        let successfullySynced = !(isInternalServerError || isNotFoundError)
+        resultDict[Backend.RCSuccessfullySyncedKey as String] = successfullySynced
+
+        let attributesResponse = (response[Backend.RCAttributeErrorsResponseKey] as? [String: Any]) ?? response
+        resultDict[Backend.RCAttributeErrorsKey] = attributesResponse[Backend.RCAttributeErrorsKey]
+
+        return resultDict
+    }
+
+    func maybeNumberFromError(code: Any?) -> NSNumber? {
+        // The code can be a String or NSNumber
+        if let codeString = code as? String {
+            if let codeInt = Int(codeString) {
+                return codeInt as NSNumber
+            } else {
+                return nil
+            }
+        }
+
+        return code as? NSNumber
+    }
+
+    // MARK: Callback cache management
+
+    func add(callback: @escaping BackendPurchaserInfoResponseHandler, key: String) -> CallbackCacheStatus {
+        return callbackQueue.sync { [self] in
+            var callbacksForKey = purchaserInfoCallbacksCache[key] ?? []
+            let requestAlreadyInFlight: CallbackCacheStatus = !callbacksForKey.isEmpty
```

Comment:
maybe this `let` should now be called cacheStatus?

Reply:
💯